tests/ci/azureml_tests/aml_utils.py

```diff
@@ -0,0 +1,199 @@
+"""
+import pathlib
+import tempfile
+
+from azure.ai.ml import MLClient, command
+from azure.ai.ml.entities import AmlCompute, BuildContext, Environment, Workspace
+from azure.ai.ml.exceptions import JobException
+from azure.core.exceptions import ResourceExistsError
+from azure.identity import DefaultAzureCredential
+
+def get_client(subscription_id, resource_group, workspace_name):
+    """
+    Get the client with specified AzureML workspace, or create one if not existing.
+    See https://github.com/Azure/azureml-examples/blob/main/sdk/python/resources/workspace/workspace.ipynb
+    """
+    params = dict(
+        credential=DefaultAzureCredential(),
+        subscription_id=subscription_id,
+        resource_group_name=resource_group,
+    )
+    client = MLClient(**params)
+
+    workspace = client.workspaces.get(workspace_name)
+    if workspace is None:
+        workspace = client.workspaces.begin_create(
+            Workspace(name=workspace_name)
+        ).result()
+
+    params["workspace_name"] = workspace_name
+    client = MLClient(**params)
+    return client
+
+
+def create_or_start_compute(client, name, size, max_instances):
+    """
+    Start the specified compute.
+    See https://github.com/Azure/azureml-examples/blob/main/sdk/python/resources/compute/compute.ipynb
+    """
+    compute = client.compute.get(name)
+    if compute is None:
+        compute = client.compute.begin_create_or_update(
+            AmlCompute(
+                name=name,
+                type="amlcompute",
+                size=size,
+                max_instances=max_instances,
+            )
+        ).result()
+
+
+def get_or_create_environment(
+    client,
+    environment_name,
+    use_gpu,
+    use_spark,
+    conda_pkg_jdk,
+    python_version,
+    commit_sha,
+):
+    """
+    AzureML requires the run environment to be setup prior to submission.
+    This configures a docker persistent compute.
+    See https://github.com/Azure/azureml-examples/blob/main/sdk/python/assets/environment/environment.ipynb
+
+    Args:
+        client (MLClient): the client to interact with AzureML services
+        environment_name (str): Environment name
+        use_gpu (bool): True if gpu packages should be
+            added to the conda environment, else False
+        use_spark (bool): True if PySpark packages should be
+            added to the conda environment, else False
+        conda_pkg_jdk (str): "openjdk=8" by default
+        python_version (str): python version, such as "3.9"
+        commit_sha (str): the commit that triggers the workflow
+    """
+    conda_env_name = "reco"
+    conda_env_yml = "environment.yml"
+    condafile = fr"""
+name: {conda_env_name}
+channels:
+  - conda-forge
+dependencies:
+  - python={python_version}
+  - {conda_pkg_jdk}
+  - pip:
+    - pymanopt@https://github.com/pymanopt/pymanopt/archive/fb36a272cdeecb21992cfd9271eb82baafeb316d.zip
```

Comment:
We have moved `pymanopt` to `extras`, so I think you don't need this line.

Reply:
I am not quit sure.  But I think the line has something to do with https://github.com/recommenders-team/recommenders/issues/2038

Reply:
Ah yes, GeoIMC has not been updated to use a recent pymanopt version. So we have this workaround for testing. You can keep it as is.